### ISSUE TITLE
`aws_ecs_service.capacity_provider_strategy` is not parsed as FARGATE deployment type

### ISSUE THREAD

**Issue by fraserc182** [infracost/infracost#1349]
I am deploying a service and the prices that infracost is giving out seem wrong.
The main issue is it seems to think that the `aws_ecs_service` is free. Am I doing something wrong here or expecting something that cannot be done?
The resource is listed as a paid service here - https://www.infracost.io/docs/supported_resources/aws/

---

*additional context added after investigation*

Infracost uses `aws_ecs_service.launch_type` to determine what type of deployment the ecs service is (EC2/FARGATE). EC2 deployments are prices through `auto_scaling_group` and `aws_ecs_service` is marked as free.

When `aws_ecs_service.launch_type` is not specified it defaults to EC2 even if `capacity_provider_strategy` is set to a FARGATE type strategy.

We should change the `aws_ecs_service` to check `capacity_provider_strategy` block before determining the deployment type

**Comment by hugorut** [comment]
hey @fraserc182 thanks for reporting this. You are correct, Infracost does support `aws_ecs_service` as a paid resource, so there must be something funky going on here! Do you mind sharing the terraform resource block that is being reported as free? This will help me narrow down the problem.

**Comment by fraserc182** [comment]
Hi @hugorut thanks for looking at this for me.
The resource in question is:

```
resource "aws_ecs_service" "django" {

  name                               = local.awsResourceName
  cluster                            = data.terraform_remote_state.cluster.outputs.ecs_clusters[var.serviceConfig.infrastructure.EcsClusterName].arn
  task_definition                    = aws_ecs_task_definition.django.arn
  desired_count                      = var.serviceConfig.infrastructure.TaskCount

  deployment_minimum_healthy_percent = 50
  deployment_maximum_percent = 200

  capacity_provider_strategy {
    capacity_provider = "FARGATE"
    weight = 0
    base = 0
  }

  capacity_provider_strategy {
    capacity_provider = "FARGATE_SPOT"
    weight = 1
  }

  network_configuration {
    subnets = [
      data.terraform_remote_state.cluster.outputs.private_az1_subnet.id,
      data.terraform_remote_state.cluster.outputs.private_az2_subnet.id,
      data.terraform_remote_state.cluster.outputs.private_az3_subnet.id
    ]
    security_groups = [aws_security_group.django.id]
  }

  platform_version = "1.4.0"

  load_balancer {
    target_group_arn = aws_lb_target_group.django.arn
    container_name   = "django"
    container_port   = 8000
  }


  tags = {
    continuous_deployment_enabled = var.continuousDeploymentEnabled
  }
  
  service_registries {
    registry_arn = aws_service_discovery_service.django.arn
    container_name   = "django"
  }

}
```

If you need more info of what the variables refer to, please let me know and I can get it.

**Comment by fraserc182** [comment]
Also, the output I get from running infracost is:

```
 OVERALL TOTAL                                                                                                  $0.30
──────────────────────────────────
28 cloud resources were detected:
∙ 4 were estimated, 1 includes usage-based costs, see https://infracost.io/usage-file
∙ 24 were free:
  ∙ 5 x aws_iam_role_policy_attachment
  ∙ 3 x aws_cloudwatch_event_rule
  ∙ 3 x aws_cloudwatch_event_target
  ∙ 3 x aws_lambda_permission
  ∙ 1 x aws_api_gateway_resource
  ∙ 1 x aws_ecs_service
  ∙ 1 x aws_ecs_task_definition
  ∙ 1 x aws_iam_role
  ∙ 1 x aws_iam_role_policy
  ∙ 1 x aws_lb_listener_rule
  ∙ 1 x aws_lb_target_group
  ∙ 1 x aws_security_group
  ∙ 1 x aws_service_discovery_service
  ∙ 1 x aws_wafv2_ip_set
```

**Comment by hugorut** [comment]
Thanks @fraserc182, let me delve into this a bit deeper and get back to you ASAP

**Comment by hugorut** [comment]
@fraserc182 - a quick clarification on your `aws_ecs_service`. Are you running this on FARGATE or EC2? From reading the spec you've given me I believe you're running on EC2, but wanted to check.

**Comment by fraserc182** [comment]
@hugorut It's fargate it's running on. Specifically the `FARGATE_SPOT` capacity provider.

**Comment by fraserc182** [comment]
@hugorut Looks much better now!
Is there any way to have infracost pick this up without adding in the `launch_type = "FARGATE"` option? 
Also, my last question is how does infracost keep track of cost increase/decrease with each plan? If there is docs for this just point me at those!
Thanks for your help, much appreciated.

```
##### Infracost estimate #####



Monthly cost will increase by $144 (+48053%%)



Previous monthly cost: $0.30

New monthly cost: $144



Infracost output:



    Project: xxx

    

    + aws_ecs_service.django

      +$144

    

        + Per GB per hour

          +$25.96

    

        + Per vCPU per hour

          +$118

    

    + aws_route53_record.django

      Monthly cost depends on usage

    

        + Standard queries (first 1B)

          Monthly cost depends on usage

            +$0.40 per 1M queries

    

        + Latency based routing queries (first 1B)

          Monthly cost depends on usage

            +$0.60 per 1M queries

    

        + Geo DNS queries (first 1B)

          Monthly cost depends on usage

            +$0.70 per 1M queries

    

    Monthly cost change for xxx

    Amount:  +$144 ($0.30 → $144)

    Percent: +48,053%

    

    ──────────────────────────────────

    Key: ~ changed, + added, - removed

    

    28 cloud resources were detected:

    ∙ 5 were estimated, 1 includes usage-based costs, see https://infracost.io/usage-file[]
    ∙ 23 were free:

      ∙ 5 x aws_iam_role_policy_attachment

      ∙ 3 x aws_cloudwatch_event_rule

      ∙ 3 x aws_cloudwatch_event_target

      ∙ 3 x aws_lambda_permission

      ∙ 1 x aws_api_gateway_resource

      ∙ 1 x aws_ecs_task_definition

      ∙ 1 x aws_iam_role

      ∙ 1 x aws_iam_role_policy

      ∙ 1 x aws_lb_listener_rule

      ∙ 1 x aws_lb_target_group

      ∙ 1 x aws_security_group

      ∙ 1 x aws_service_discovery_service

      ∙ 1 x aws_wafv2_ip_set

    

    Add cost estimates to your pull requests: https://infracost.io/cicd
```
```

**Comment by alikhajeh1** [comment]
>  Also, my last question is how does infracost keep track of cost increase/decrease with each plan? If there is docs for this just point me at those!

We don't have that, the closest thing we have is https://www.infracost.io/docs/features/share_links/ that can be used to generate a links for specific snapshots (e.g. you can take that at the start of the week, and compare it with the one next week manually). Out of interest, why do you want this? (the context helps us a lot)

**Comment by fraserc182** [comment]
Ah I was referring to this in the output:

I think I get it now, it's not keeping track of our costs. It just adds or subtracts based on the plan. That makes more sense!

```
Monthly cost will increase by $0.00 (+0%%)


Previous monthly cost: $230

New monthly cost: $230
```

**Comment by hugorut** [comment]
@fraserc182 - good to hear this fixed your issue. We can definitely add the option in to infer the launch type from child blocks in the resource. The current issue is that `launch_type` defaults to `EC2` in the aws provider and this isn't overwritten by the `capacity_provider_strategy` option. So we'll need to add some additional code to make sure that the child blocks aren't specifying FARGATE. Shouldn't be too much work, but in the meantime i'd recommend just adding the `launch_type` attribute as this shouldn't change your deployment at all.

> Also, my last question is how does infracost keep track of cost increase/decrease with each plan? If there is docs for this just point me at those!

Regarding this and building on what @alikhajeh1 has mentioned, we don't have a way for you to view historic fluctuations of prices across infracost output. However, infracost infers increases/decreases from individual plans by parsing Terraform plan changes working out prices from the "changed resources"

**Comment by hugorut** [comment]
I'm going to change the title and description of this issue so that someone can pick it up to work on soon.

**Comment by fraserc182** [comment]
Perfect, thanks for the help with this. Appreciate you picking it up so quickly as well.

**Comment by hugorut** [comment]
Hey @fraserc182, this is now fixed in the [0.9.22](https://github.com/infracost/infracost/releases/tag/v0.9.22) release. Thanks for reporting this issue! 👍
